# COGS 108 - Data Checkpoint

## Authors

TEAM LIST AND CREDITS

Eljin Vedar
Writing – original draft, Writing –  review & editing 

Isaac Chen
Writing – original draft, Writing –  review & editing

Jiafan Du
Writing – original draft, Writing –  review & editing

Robin Villareal
Writing – original draft, Writing –  review & editing 

Barry Tunstall
Writing – original draft, Writing –  review & editing 


## Research Question

“How well does performance on a Dance Dance Revolution chart with difficulty X, defined by the game’s numeric rating, predict a player’s overall skill level across songs of varying difficulty?”


For this idea, we aim to address the new player approach to finding a difficulty range. For our rhythm game of interest, we will choose Dance Dance Revolution (DDR), a classic rhythm game. Starting out, it may feel like the game is too easy or too hard. Given pre-existing game stats, we will analyze the data given by a player to make a prediction of their current skill range, or what difficulty X number a player can attempt on average. Furthermore, with this prediction, we can go further into recommending new songs for a player to try out relative to their corresponding difficulty X.


## Background and Prior Work

As members of the Dance and Rhythm Game Club, we stand to address socioeconomic factors by making arcade rhythm games accessible. We understand the constraints of accessibility to an arcade, time, and disposable income play in being able to experience arcade rhythm games like Dance Dance Revolution (DDR) . It is one of our motivations to implore the general public to try DDR as we bring the arcade experience to campus. As we seek to expand our community, one of the problems we face when reaching out to new players is the willingness to try the game i.e not knowing where to begin. We run into phrases such as “I’m not good”, “I’ve never played before”, and even “I don’t even know where to start!” constantly. To that, we want to devise a method that involves a beginner friendly approach; easing players into DDR assuming they have no experience but an eagerness to learn. For the project, we used three sources based on their relevance to predictive analysis to form the foundation of our research.

 The paper, “Dance Dance Convolution,” [3](#ref3) is a study that attempts to generate DDR step charts based on machine learning, with two inputs - an audio file, and a desired difficulty. The model would then generate a fully playable chart by decomposing the task into two sequential prediction problems: first determining when steps should occur, and then determining which arrows should be placed at each predicted timing. The results were analyzed based on how similar they were to the original chart - things like average hold note length, arrow selection, and beat placement. The important conclusions that come from the study was the capacity to assign chart difficulty ratings for procedurally generated charts based on note densities, although limited to a max bpm of 120; expanded on in a revisit in 2025 [1](#ref1) which includes a refined version exceeding the 120 bpm threshold. Another useful prediction scheme was being able to predict the chart difficulty [2](#ref2) based on attributes we have assigned as step data and technique data. 
The main takeaway from the results of the paper, for our uses, is the conclusion that DDR charting aligns closely with underlying musical structure, and that difficulty can be linearly correlated with step frequency, rhythmic density, and other computationally extractable metrics. This is useful for our predictive modeling because it suggests that the player performance data we gather, such as timing accuracy, hit‑window distribution, and error frequency can also be correlated with chart difficulty in a uniform way. Thus, the same mathematical relationships that allow systems like DDC to infer difficulty from audio and chart context can be inverted in our analysis - we can infer a player's effective skill level by measuring how their accuracy degrades or holds steady across segments of varying structural difficulty.

The “Dance Dance ConvLSTM” [1](#ref1) paper is an extension of this paper, diving deeper into subsections of charting in order to further quantify patterns that arise in step charting - correlated to chart triggers such as music speedups, slowdowns, or certain step patterns, such as repeated notes on the same arrow - these are usually considered to be more difficult to execute than standard patterning. [2](#ref2) This provides a further foundation for our analysis: if machine‑learning models can reliably infer difficulty from musical and structural cues, then player‑performance data should likewise reflect predictable responses to those same cues, enabling skill prediction through our structural‑difficulty analysis.

The third paper, “Predicting Chart Difficulty in Rhythm Games Through Classification Using Chart Pattern Derived Attributes,” also deals with the subject of correlating the in-game difficulty numbers with actual difficulty. Their analysis deals with difficulty and correlating it with chart structure - The main conclusions from the paper, useful for our purposes, is that objective structural difficulty can be quantified and measured, and that the charting in the game’s maps is correlated with this difficulty. [3](#ref3) This is useful for us because it means that we can apply a similar analysis to player skill, since performance data (such as timing accuracy, error rate, and stamina degradation, etc.) should correlate with these same difficulty signifiers in the charts. Essentially, If difficulty can be reliably inferred from chart patterns, then player capability can be inferred from how well a player handles those patterns.

From the articles and our own playing experience, we found 3 distinct data categories relevant to our project:
1. Song Data (exact song) - data showing the title, artist, and folder the song is located in. This is important because there are versions of DDR that run on a different difficulty scale; making sure the songs user play are in the arcade version of the game.

2. Chart Data (Difficulty X) - The degree of gameplay a player can encounter during a song. Individual songs in DDR can be played on varying difficulty levels. Each difficulty iteration is commonly referred to as a chart. Charts are given numerical values assigned by DDR from 1-20 (we will assign this number to X). As the number increases, players can expect an increase of notes on the screen (note density) and rhythmic patterns. Charts of equal value have the tendency to feature similar complexity. For example, Song A with difficulty 5 will be as equal as Song B with difficulty 5 in terms of chart complexity.

3. Performance data (how a player does) - the overall score an individual obtains along with a timeline of where the player missed notes. This works in tandem with step data and technique data to evaluate what core fundamentals the person has/needs improvement on. In our project, we will use some of DDR's iconic scoring systems to assess new players (more info in HYPOTHESIS)



Although these articles make great strides in the development of generated difficulty, we find the limitation of these contexts to engage with the new player community and the base game of DDR. Having the opportunity to expand on these works, we want to take chart predictions and the data acquired to contextualize them for the use of a new player; giving them a direction as they start their rhythm game journey.

1. <a name="ref1"> </a> O’Malley, Miguel. Dance Dance ConvLSTM. 2025, arXiv:2507.01644. https://arxiv.org/pdf/2507.01644
2. <a name="ref2"> </a>Caronongan, Arturo P., III, and Nelson A. Marcos. “Predicting Chart Difficulty in Rhythm Games Through Classification Using Chart Pattern Derived Attributes.” Computational Science and Technology, 2021, pp. 193–205. Springer Nature. https://link.springer.com/chapter/10.1007/978-981-33-4069-5_17
3. <a name="ref3"> </a>Donahue, Chris, Zachary C. Lipton, and Julian McAuley. “Dance Dance Convolution.” Proceedings of the 34th International Conference on Machine Learning, PMLR 70, 2017. https://proceedings.mlr.press/v70/donahue17a/donahue17a.pdf

## Hypothesis


As players take on higher difficulty X ratings, DDR charts demand greater stamina, precision, and pattern recognition due to chart complexity increasing (i.e total stepcount, rhythms implemented, how many notes appear at once). Therefore, players who achieve EX% around the community avg EX% reflect the capacity to comprehend and execute these factors. We come to this conclusion since chart complexity can be generalized across songs of equal difficulty X rating; reflecting their underlying skill. 


We predict a strong correlation between a player's EX% reflecting their overall skill level when compared to the avg EX%.


To analyze the skill of a player, we have used/adjusted preexisting variables DDR currently implements to assess players:
1. Difficulty X - As stated previously, DDR difficulty has a rating numbered from easiest to hardest: 1-20. DDR includes 4 standard labels: BEGINNER 1-4 , BASIC 3-7, DIFFICULT 5-9, EXPERT 10-13+. One issue that has arisen was the overlapping ranges in the standard labels. As an example, one song can have a BASIC 7, while another song can have a DIFFICULT 7 even though there is no major difficulty spike between the two. For a new player, seeing DIFFICULT 7 instead of BASIC 7 may afflict a bias when choosing the difficulty. To avoid ambiguity, we want to simplify the system purely on the numeric RATING; assigning it to our variable, X (X can range from 1-20). When we refer to difficulty X now, it provides a clearer representation of difficulty while still utilizing the pre-existing DDR system. (a 7 is a 7!).


2. EX - EX is a scoring system originating from DDR EXTREME; independent from the in game combo and life bar players use. EX was designed to be a timing-based accuracy score, given by the summation of the 2 best timing judgements - Marvelous and Perfect. To calculate EX, the formula is given below:
        
        EX = ∑ 2(Marvelous) + 1(Perfect)



3. EX% - EX% is EX normalized. Because charts have varying total notecounts EX alone is
Not enough to determine overall skill level. By normalizing EX, we obtain a metric that allows us to compare charts of different total notecounts. To calculate EX% we use the formula:
        
        EX% = EX/(3 x Total Notecount)


4. median EX% (med EX%)  - The median EX% on a given song that will give us the baseline for a player's performance in relation to players who also attempted  chart difficulty X. We understand that one of the constraints of community-driven data is the possibility of inflated averages due to the data coming from registered users. These registered users offer a higher level of time commitment and scoring data than our new player demographic. For our project Median EX% is used to operationalize target performance that a player must achieve to be considered for that difficulty X skill level.


5. Overall skill level - A player's overall skill level is operationalized as their individual EX% on a difficulty X chart in relation to the given difficulty X’s median EX%. This is best represented by the difficulty X they can reliably play.



## Data

### Data overview
- Dataset #1
  - Dataset Name: DDR BPI / Skill Attack data
  - Link to the dataset: https://github.com/COGS108/Group051_WI26/blob/64df21f9fb6148e4e860e4b8c466c7080442b57d/data/00-raw/DDR%20BPI%20_%20Skill%20Attack%20data%20-%20bpi.csv

    Dataset as Spreadsheet: https://docs.google.com/spreadsheets/d/1_vvxS71takEt3gIjkC2vcrI6sHNfTwWKtSoETwBmY5E/edit?gid=660287600#gid=660287600
  
  - Number of observations: 5894

  - Number of variables: 15

  - Description of the variables most relevant to this project

    1. DIFFICULTY X - Numeric difficulty of a song

    2. Max EX - Maximum possible EX score on a given chart

    3. PLAYERS - How many players attempted the given chart

  - Descriptions of any shortcomings this dataset has with respect to the project

    The dataset includes songs outside the arcade version of DDR. Although songs outside the arcade version may adopt a difficulty X, there IS a noticeable spike in difficulty compared to an arcade song counterpart, so we seek to remove them from our dataset. Including them may also have an affect on our median EX%.

- Dataset #2
  - Dataset Name: Single (Chart Analyzer)
  - Link to the dataset: https://github.com/COGS108/Group051_WI26/blob/64df21f9fb6148e4e860e4b8c466c7080442b57d/data/00-raw/DanceDanceRevolution%20Chart%20Data%20Spreadsheet%20-%20Single%20(Chart%20Analyzer).csv

  Dataset as Spreadsheet: https://docs.google.com/spreadsheets/d/10NT1VPYV8JHEpi68SBCTApOO5mWEjIAbJP3C5Csthyk/edit?gid=2049250718#gid=2049250718
  - Number of observations: 12866

  - Number of variables: 260
  - Description of the variables most relevant to this project

    1. AC - A song is featured in the arcade version of the game
    
    2. INTL - the song is present in ALL regions an arcade version exists.


  - Descriptions of any shortcomings this dataset has with respect to the project

    Most of the relevant data overlaps with dataset 1 but is even more skewed because the averages are in tandem with all songs of a corresponding difficulty X, even if they are outside the arcade version of the game.

- Dataset #3
  - Dataset Name: Animefest Player Scores
  - Link to the dataset: https://github.com/COGS108/Group051_WI26/blob/64df21f9fb6148e4e860e4b8c466c7080442b57d/data/00-raw/parsed.csv
  - Number of observations: 198
  - Number of variables: 49

  - Description of the variables most relevant to this project
    1. Title - What song they played
    2. Difficulty - uses their beginner/basic/difficult/expert labels
    3.  Score - EX scores are given as EX x 1,000,000

  - Descriptions of any shortcomings this dataset has with repsect to the project
  
    Because the dataset uses the beginner/basic/difficult/expert labels we will have to search through the other sets for their numeric ratings. Another concern is the EX score is scaled by 1,000,000 so all EX scores from this dataset must be divided by 1,000,000 before it can be comparable to the averages.







In [1]:
# Run this code every time when you're actively developing modules in .py files.  It's not needed if you aren't making modules
#
## this code is necessary for making sure that any modules we load are updated here 
## when their source code .py files are modified

%load_ext autoreload
%autoreload 2

In [2]:
# Setup code -- this only needs to be run once after cloning the repo!
# this code downloads the data from its source to the `data/00-raw/` directory
# if the data hasn't updated you don't need to do this again!

# if you don't already have these packages (you should!) uncomment this line
# %pip install requests tqdm

import sys
sys.path.append('./modules') # this tells python where to look for modules to import

import get_data # this is where we get the function we need to download data

# replace the urls and filenames in this list with your actual datafiles
# yes you can use Google drive share links or whatever
# format is a list of dictionaries; 
# each dict has keys of 
#   'url' where the resource is located
#   'filename' for the local filename where it will be stored 
datafiles = [
    { 'url': 'https://raw.githubusercontent.com/COGS108/Group051_WI26/64df21f9fb6148e4e860e4b8c466c7080442b57d/data/00-raw/DDR%20BPI%20_%20Skill%20Attack%20data%20-%20bpi.csv', 'filename':'DDR BPI _ Skill Attack data - bpi.csv'},
    { 'url': 'https://raw.githubusercontent.com/COGS108/Group051_WI26/64df21f9fb6148e4e860e4b8c466c7080442b57d/data/00-raw/DanceDanceRevolution%20Chart%20Data%20Spreadsheet%20-%20Single%20(Chart%20Analyzer).csv', 'filename':'DanceDanceRevolution Chart Data Spreadsheet - Single (Chart Analyzer).csv'},
    { 'url': 'https://raw.githubusercontent.com/COGS108/Group051_WI26/refs/heads/master/data/00-raw/parsed.csv', 'filename':'parsed.csv' }
]

get_data.get_raw(datafiles,destination_directory='data/00-raw/')

Overall Download Progress:  33%|███▎      | 1/3 [00:00<00:00,  3.35it/s]                     

Successfully downloaded: DDR BPI _ Skill Attack data - bpi.csv



Overall Download Progress:  67%|██████▋   | 2/3 [00:03<00:01,  1.72s/it]                                       

Successfully downloaded: DanceDanceRevolution Chart Data Spreadsheet - Single (Chart Analyzer).csv



Overall Download Progress: 100%|██████████| 3/3 [00:03<00:00,  1.07s/it]

Successfully downloaded: parsed.csv


### DDR BPI/Skill Attack Performance

The DDR BPI/Skill Attack Performance Dataset is a community-driven dataset that catalogs performance based statistics given by registered users who uploaded their playing data. Each row corresponds to a specific song, difficulty level/rating, and provides important statistics such as the average EX score obtained along and how many players were involved to produce those averages.
The relevant columns from the dataset are as follows:

diff - Numeric difficulty of a song. Can take a value between 1-20

Max EX - Maximum possible EX score on a given chart

PLAYERS - How many players attempted the given chart

A central metric in this dataset is the EX score, which is the accuracy‑based scoring system that is the foundation of our project (EX Formula). These EX’s are important to log when we start converting them into EX%’s.

One major concern of the dataset is how it is community-driven by registered users. We are building our averages on self-selected players who are willing to upload their scoring data for review. This may lead to introducing our target audience of newer players to the standards of the competitive scene. Through statistical analysis, we will attempt to make newer players comparable with some of the expected high averages via computing median EX% instead.

3. Use the cell below to 
    1. load the dataset 
    2. make the dataset tidy or demonstrate that it was already tidy
    3. demonstrate the size of the dataset
    4. find out how much data is missing, where its missing, and if its missing at random or seems to have any systematic relationships in its missingness
    5. find and flag any outliers or suspicious entries
    6. clean the data or demonstrate that it was already clean.  You may choose how to deal with missingness (dropna of fillna... how='any' or 'all') and you should justify your choice in some way
    7. You will load raw data from `data/00-raw/`, you will (optionally) write intermediate stages of your work to `data/01-interim` and you will write the final fully wrangled version of your data to `data/02-processed`
4. Optionally you can also show some summary statistics for variables that you think are important to the project
5. Feel free to add more cells here if that's helpful for you


In [3]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
import pandas as pd
import numpy as np
from scipy import stats
Dataset1 = pd.read_csv('data/00-raw/DDR BPI _ Skill Attack data - bpi.csv') # Read the raw data into a dataframe
display(Dataset1)

title      id chart  90% EX bpi  avg EX  90% EX  \
0                          ENDYMION   756.0   CDP      105.76   525.0  2511.0   
1               BREAKING THE FUTURE  1188.0   CDP      105.22   724.0  2443.0   
2                        EGOISM 440   640.0   CDP      104.70   369.0  2157.0   
3               BREAKING THE FUTURE  1188.0   CSP      104.22  1071.0  2605.0   
4     Over The &ldquo;Period&rdquo;   648.0   CDP      103.90   546.0  2116.0   
...                             ...     ...   ...         ...     ...     ...   
9612                 MAKE IT BETTER     NaN   NaN         NaN     NaN     NaN   
9613                 MAKE IT BETTER     NaN   NaN         NaN     NaN     NaN   
9614                 MAKE IT BETTER     NaN   NaN         NaN     NaN     NaN   
9615                 MAKE IT BETTER     NaN   NaN         NaN     NaN     NaN   
9616                 MAKE IT BETTER     NaN   NaN         NaN     NaN     NaN   

       wr EX  max EX  diff  avg money  wr money  players  mfcs  notecount  \
0     1944.0  2790.0  19.0   285278.0  870250.0    286.0   0.0      930.0   
1     1959.0  2715.0  19.0   400197.0  886950.0     51.0   0.0      905.0   
2     1792.0  2397.0  19.0   249550.0  908150.0    410.0   0.0      799.0   
3     2197.0  2895.0  19.0   544822.0  911730.0    281.0   0.0      965.0   
4     1827.0  2352.0  19.0   354282.0  924670.0    333.0   0.0      784.0   
...      ...     ...   ...        ...       ...      ...   ...        ...   
9612     NaN     NaN   NaN        NaN       NaN      NaN   NaN        NaN   
9613     NaN     NaN   NaN        NaN       NaN      NaN   NaN        NaN   
9614     NaN     NaN   NaN        NaN       NaN      NaN   NaN        NaN   
9615     NaN     NaN   NaN        NaN       NaN      NaN   NaN        NaN   
9616     NaN     NaN   NaN        NaN       NaN      NaN   NaN        NaN   

      untimed  
0        30.0  
1        38.0  
2         1.0  
3        41.0  
4        25.0  
...       ...  
9612      NaN  
9613      NaN  
9614      NaN  
9615      NaN  
9616      NaN  

[9617 rows x 15 columns]

As seen above this data is already tidy and in a usable format. The dataset looks tidy on the surface because it has consistent column names, each row represents one chart, numeric columns are mostly numeric, there are no obvious formatting errors, and the structure is already tidy (one row per chart, one column per variable). The dataset has 9617 Entries and 15 columns of data. Next we'll clean the data and find where and how data is missing.

In [4]:
Dataset1[Dataset1.isnull().any(axis=1)] # Display rows with a null value

title  id chart  90% EX bpi  avg EX  90% EX  wr EX  max EX  \
9116  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9117  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9118  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9119  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9120  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
...              ...  ..   ...         ...     ...     ...    ...     ...   
9612  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9613  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9614  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9615  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   
9616  MAKE IT BETTER NaN   NaN         NaN     NaN     NaN    NaN     NaN   

      diff  avg money  wr money  players  mfcs  notecount  untimed  
9116   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9117   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9118   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9119   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9120   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
...    ...        ...       ...      ...   ...        ...      ...  
9612   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9613   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9614   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9615   NaN        NaN       NaN      NaN   NaN        NaN      NaN  
9616   NaN        NaN       NaN      NaN   NaN        NaN      NaN  

[501 rows x 15 columns]

All missing data contained in this dataset is on rows 9116-9616 (501) and they all contain no variables thus they can be dropped for the purposes of this project.

In [5]:
Dataset1 = Dataset1.dropna()
Dataset1

title      id  chart  90% EX bpi       avg EX  \
0                          ENDYMION   756.0    CDP      105.76   525.000000   
1               BREAKING THE FUTURE  1188.0    CDP      105.22   724.000000   
2                        EGOISM 440   640.0    CDP      104.70   369.000000   
3               BREAKING THE FUTURE  1188.0    CSP      104.22  1071.000000   
4     Over The &ldquo;Period&rdquo;   648.0    CDP      103.90   546.000000   
...                             ...     ...    ...         ...          ...   
9111                   SAY A PRAYER  1191.0  begSP        -inf   129.281107   
9112                Chromatic Burst  1199.0  begSP        -inf   125.816852   
9113                      千客万来☆無問題！  1244.0  begSP        -inf   261.326531   
9114                        走れメロンパン  1247.0  begSP        -inf   156.931818   
9115                  melody H4CKER  1260.0  begSP        -inf   191.804002   

      90% EX   wr EX  max EX  diff  avg money   wr money  players  mfcs  \
0     2511.0  1944.0  2790.0  19.0   285278.0   870250.0    286.0   0.0   
1     2443.0  1959.0  2715.0  19.0   400197.0   886950.0     51.0   0.0   
2     2157.0  1792.0  2397.0  19.0   249550.0   908150.0    410.0   0.0   
3     2605.0  2197.0  2895.0  19.0   544822.0   911730.0    281.0   0.0   
4     2116.0  1827.0  2352.0  19.0   354282.0   924670.0    333.0   0.0   
...      ...     ...     ...   ...        ...        ...      ...   ...   
9111   126.0   141.0   141.0   2.0   993675.0  1000000.0     46.0  24.0   
9112   124.0   138.0   138.0   1.0   992087.0  1000000.0     65.0  34.0   
9113   245.0   273.0   273.0   3.0   999883.0  1000000.0     49.0  25.0   
9114   143.0   159.0   159.0   2.0   999979.0  1000000.0     44.0  22.0   
9115   180.0   201.0   201.0   2.0   998942.0  1000000.0     26.0  13.0   

      notecount  untimed  
0         930.0     30.0  
1         905.0     38.0  
2         799.0      1.0  
3         965.0     41.0  
4         784.0     25.0  
...         ...      ...  
9111       47.0      0.0  
9112       46.0      0.0  
9113       91.0      0.0  
9114       53.0      0.0  
9115       67.0      4.0  

[9116 rows x 15 columns]

After dropping missing rows we are left with 9116 Entries in this dataset. Next we'll search for outliers and suspicious entries

In [6]:
Dataset1.describe()

id   90% EX bpi       avg EX       90% EX        wr EX  \
count  9116.000000  9116.000000  9116.000000  9116.000000  9116.000000   
mean    636.326349         -inf   753.763501   828.215774   918.483107   
std     359.100714          NaN   354.950138   442.108494   486.212369   
min       0.000000         -inf    63.747588    62.000000    69.000000   
25%     339.750000     0.057804   479.157693   493.250000   548.250000   
50%     632.500000     0.730000   709.630595   750.000000   834.000000   
75%     944.000000     2.500000   983.000000  1074.000000  1191.000000   
max    1260.000000   105.760000  2118.326556  2867.000000  3016.000000   

            max EX         diff      avg money        wr money      players  \
count  9116.000000  9116.000000    9116.000000     9116.000000  9116.000000   
mean    920.738372     9.092804  958811.723892   999874.542563   552.164765   
std     491.227426     4.171732   46409.852708     2679.951383   537.368528   
min      69.000000     1.000000  249550.000000   870250.000000    21.000000   
25%     548.250000     6.000000  950866.250000  1000000.000000   187.000000   
50%     834.000000     9.000000  971481.500000  1000000.000000   337.000000   
75%    1194.000000    12.000000  983670.250000  1000000.000000   699.000000   
max    3186.000000    19.000000  999990.000000  1000000.000000  2882.000000   

              mfcs    notecount      untimed  
count  9116.000000  9116.000000  9116.000000  
mean     19.061540   306.912791    14.617376  
std      22.706698   163.742475    16.751520  
min       0.000000    23.000000     0.000000  
25%       2.000000   182.750000     4.000000  
50%      10.000000   278.000000    11.000000  
75%      29.000000   398.000000    20.000000  
max     399.000000  1062.000000   381.000000

Based on these results if we choose to work with '90% EX bpi' we'll need to be careful to work around the -inf values. Other than this no glaring outliers are instantly visible. Finally we can input the dataframe into our interim data folder.

In [7]:
Dataset1.to_csv(r'data/01-interim/bpi_interim.csv', index=False)

### Data Cleaning: DDR BPI/Skill Attack Performance
For Dataset 1, the raw table was already tidy (one row per observation, one column per variable). The only cleaning needed was removing a small set of systematically incomplete rows with missing values.

For dataset 1 all that is left for us is to eliminate columns and rows that are unnecessary for our project purpose such as the excess variables that we did not see as important or songs that are not part of the arcade version of DDR (which we will cross reference once we combine the data used from dataset 2).


### Single (Chart Analyzer) 
Single (chart analyzer) provides a similar approach to DDR BPI/Skill Attack Performance but integrates performance data from DDR GRAND PRIX and online version of DDR that may include custom songs or modified difficulties. Because of this, we can not use the provided EX scores. The main reason we included this dataset was because Single (chart analyzer) has a more in-depth approach to song titles. Song titles in this dataset are labeled by their presence in the DDR scene - giving us an indicator if a song is available in an arcade version of DDR. The 2 main variables we must pay attention to are the following:

AC - arcade cabinet

INTL - international

By finding songs with these labels, we can conclude if the song is 1. In the arcade version an 2. Accessible to anyone no matter the arcade. By crossreferencing this dataset with DDR BPI/Skill Attack Performance, we make our personal song encyclopedia even stronger.
Limitations were already expressed through the statistics being invalid from this dataset. The important takeaway was song verification.

In [8]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
Dataset2 = pd.read_csv('data/00-raw/DanceDanceRevolution Chart Data Spreadsheet - Single (Chart Analyzer).csv', low_memory=False) # Read raw data into DF solving for DtypeWarning
display(Dataset2)
#Dataset2.isna().sum().sum()
#Dataset2[Dataset2.isna().any(axis=1)]

Folder  \
0              DDR (AC) (JP)   
1              DDR (AC) (JP)   
2              DDR (AC) (JP)   
3              DDR (AC) (JP)   
4              DDR (AC) (JP)   
...                      ...   
12860  OHA-STA DDR (PS) (JP)   
12861  OHA-STA DDR (PS) (JP)   
12862  OHA-STA DDR (PS) (JP)   
12863  OHA-STA DDR (PS) (JP)   
12864  OHA-STA DDR (PS) (JP)   

                                                   Title Difficulty  Rating  \
0                                              Butterfly   Beginner       2   
1                                              Butterfly      Basic       5   
2                                              Butterfly  Difficult       6   
3                                              Butterfly     Expert       8   
4                                              Butterfly  Challenge      11   
...                                                  ...        ...     ...   
12860     WHY!～THE ZONA'S SONG～ | WHY! ~THE ZONA'S SONG~  Difficult       7   
12861     WHY!～THE ZONA'S SONG～ | WHY! ~THE ZONA'S SONG~     Expert       8   
12862  ゾナパラ [WHY!パラパラ・リミックス] | ZONAPARA [WHY! PARAPAR...      Basic       1   
12863  ゾナパラ [WHY!パラパラ・リミックス] | ZONAPARA [WHY! PARAPAR...  Difficult       4   
12864  ゾナパラ [WHY!パラパラ・リミックス] | ZONAPARA [WHY! PARAPAR...     Expert       6   

       Song Length  Display BPM (min)  Display BPM (max)  True BPM (min)  \
0               93                135                135           31.25   
1               93                135                135           31.25   
2               93                135                135           31.25   
3               93                135                135           31.25   
4               93                135                135           31.25   
...            ...                ...                ...             ...   
12860           86                180                180          180.08   
12861           86                180                180          180.08   
12862           99                158                158          157.98   
12863           99                158                158          157.98   
12864           99                158                158          157.98   

       True BPM (max)  BPM Changes  ...  Ambiguous Notes (H)  \
0             134.922            1  ...                    0   
1             134.922            1  ...                    4   
2             134.922            1  ...                   14   
3             134.922            1  ...                    6   
4             134.922            1  ...                    9   
...               ...          ...  ...                  ...   
12860         180.080            0  ...                    4   
12861         180.080            0  ...                    5   
12862         157.980            0  ...                    0   
12863         157.980            0  ...                    6   
12864         157.980            0  ...                    5   

      Ambiguous Notes/Step (H)  Ambiguous Notes/Second (H)  \
0                     0.000000                    0.000000   
1                     0.025000                    0.049152   
2                     0.087500                    0.172032   
3                     0.028169                    0.076661   
4                     0.027864                    0.114991   
...                        ...                         ...   
12860                 0.019048                    0.051747   
12861                 0.018797                    0.064684   
12862                 0.000000                    0.000000   
12863                 0.036364                    0.068687   
12864                 0.025773                    0.057239   

       270* Crossovers Avoided/Step (H)  270* Crossovers Avoided/Second (H)  \
0                              0.000000                            0.000000   
1                              0.000000                            0.000000   
2                       

As seen above this data is already tidy and in a usable format. The dataset looks tidy on the surface because it has consistent column names, each row represents one Chart, numeric columns are mostly numeric, there are no obvious formatting errors, and the structure is already tidy (one row per chart, one column per variable). The dataset has 12865 Entries and 260 columns of data. Next we'll clean the data and find where and how data is missing.

In [9]:
print("Sum of NaN's " + str(Dataset2.isna().sum().sum())) # Print total number of NaN's in the dataset
null_locs = (Dataset2.isna().stack().loc[lambda s: s].reset_index().rename(columns={"level_0": "row", "level_1": "column", 0: "is_null"})) # Create filter which will show all NaN row and columns

null_locs[["row", "column"]].head(25) # Display results

Sum of NaN's 23


row                                         column
0   4434  Double-step (Avoid Twist While Facing S) (DS)
1   6209  Double-step (Avoid Twist While Facing S) (DS)
2   7423  Double-step (Avoid Twist While Facing S) (DS)
3   8311  Double-step (Avoid Twist While Facing S) (DS)
4   8331  Double-step (Avoid Twist While Facing S) (DS)
5   9526  Double-step (Avoid Twist While Facing S) (DS)
6   9643  Double-step (Avoid Twist While Facing S) (DS)
7   9920                    Double-steps Into Jacks (H)
8   9920                    Double-steps Into Jumps (H)
9   9920               Double-step (Avoid Facing S) (H)
10  9920                       Ended on Double-step (H)
11  9921                    Double-steps Into Jacks (H)
12  9921                    Double-steps Into Jumps (H)
13  9921               Double-step (Avoid Facing S) (H)
14  9921                       Ended on Double-step (H)
15  9922                    Double-steps Into Jacks (H)
16  9922                    Double-steps Into Jumps (H)
17  9922               Double-step (Avoid Facing S) (H)
18  9922                       Ended on Double-step (H)
19  9923                    Double-steps Into Jacks (H)
20  9923                    Double-steps Into Jumps (H)
21  9923               Double-step (Avoid Facing S) (H)
22  9923                       Ended on Double-step (H)

We can see there are a total of 11 rows which have null values, some have multiple Null values. None of these values are in columns of interest therefore we do not need to drop these rows for our future analysis. Next we'll look for outliers.

In [10]:
df = Dataset2

# numeric-only for outlier detection
num = df.select_dtypes(include="number")

# tail cutoffs
lo = num.quantile(0.01)
hi = num.quantile(0.99)

# True where a value is in the extreme tails of its column
outlier_mask = num.lt(lo) | num.gt(hi)

# require at least k columns in the row to be flagged
k = 10

row_outlier_counts = outlier_mask.sum(axis=1)
rows_flagged = row_outlier_counts >= k

print("Total rows:", len(df))
print(f"Rows flagged (>= {k} columns extreme):", rows_flagged.sum())

# show the worst rows first (most flagged columns)
worst_idx = row_outlier_counts.sort_values(ascending=False).head(50).index

# full set of flagged rows (all columns)
df_flagged = df.loc[rows_flagged].copy()
df.loc[worst_idx].assign(outlier_flag_count=row_outlier_counts.loc[worst_idx]).head(rows_flagged.sum())

Total rows: 12865
Rows flagged (>= 10 columns extreme): 1008


Folder  \
7459                                   DDR II (Wii) (NA)   
4057                                   DDR A (AC) (INTL)   
7458                                   DDR II (Wii) (NA)   
5362                            DDR A20 PLUS (AC) (INTL)   
6121                                  DDR A3 (AC) (INTL)   
2645                           DDR ULTRAMIX4 (Xbox) (NA)   
9140                       DDR UNIVERSE2 (Xbox 360) (NA)   
8345                   DDR SuperNOVA Missions (PS2) (NA)   
8344                   DDR SuperNOVA Missions (PS2) (NA)   
6271                                  DDR A3 (AC) (INTL)   
2637                           DDR ULTRAMIX4 (Xbox) (NA)   
5584                                  DDR A3 (AC) (INTL)   
9132                       DDR UNIVERSE2 (Xbox 360) (NA)   
2641                           DDR ULTRAMIX4 (Xbox) (NA)   
7499                                   DDR II (Wii) (NA)   
6886                        DDR HOTTEST PARTY (Wii) (NA)   
7254                      DDR HOTTEST PARTY 3 (Wii) (NA)   
5826                                  DDR A3 (AC) (INTL)   
2649                           DDR ULTRAMIX4 (Xbox) (NA)   
8847                        DDR UNIVERSE (Xbox 360) (NA)   
11163                                 DDR X2 (AC) (INTL)   
8855                        DDR UNIVERSE (Xbox 360) (NA)   
10700                                  DDR X (AC) (INTL)   
6015                                  DDR A3 (AC) (INTL)   
8346                   DDR SuperNOVA Missions (PS2) (NA)   
4575                                 DDR A20 (AC) (INTL)   
2945                         DDR (PS3) (NA) Premium Pack   
8851                        DDR UNIVERSE (Xbox 360) (NA)   
7025                        DDR HOTTEST PARTY (Wii) (NA)   
9128                       DDR UNIVERSE2 (Xbox 360) (NA)   
4738                                 DDR A20 (AC) (INTL)   
6314                                  DDR A3 (AC) (INTL)   
4255                                   DDR A (AC) (INTL)   
11430                         DDR X3 VS 2ndMIX (AC) (JP)   
2653                           DDR ULTRAMIX4 (Xbox) (NA)   
9136                       DDR UNIVERSE2 (Xbox 360) (NA)   
2953                                      DDR (Wii) (NA)   
10234                              DDR WORLD (AC) (INTL)   
3703                                  DDR 2014 (AC) (JP)   
5914                                  DDR A3 (AC) (INTL)   
6926                        DDR HOTTEST PARTY (Wii) (NA)   
1601   DDR Mario Mix (GC) (NA) | DDR with MARIO (GC) ...   
6906                        DDR HOTTEST PARTY (Wii) (NA)   
9127                       DDR UNIVERSE2 (Xbox 360) (NA)   
6976                        DDR HOTTEST PARTY (Wii) (NA)   
6301                                  DDR A3 (AC) (INTL)   
3879                                  DDR 2014 (AC) (JP)   
10277                              DDR WORLD (AC) (INTL)   
3668                                  DDR 2014 (AC) (JP)   
8366                   DDR SuperNOVA Missions (PS2) (NA)   

                                                   Title     Difficulty  \
7459                                New Decade Full Song      Challenge   
4057                                            ENDYMION      Challenge   
7458                                New Decade Full Song         Expert   
5362   灼熱 Pt.2 Long Train Running | Shakunetsu Pt.2 L...      Challenge   
6121            量子の海のリントヴルム | Ryoushi no umi no Lindwurm      Challenge   
2645                              Power 3 (Industry Mix)         Expert   
9140                                 Power 5 (Aggro Mix)         Expert   
8345                                   FESTIVAL (BPM154)  (?) Difficult   
8344                                   FESTIVAL (BPM154)      (?) Basic   
6271                 トランスダンスアナーキー | Trance dance anarcky      Challenge   
2637                          Power 1 (Steppin' Out Mix)         Expert   
5584                                 BREAKING THE FUTURE      Challenge   
9132          

Based on these results we can assume the data has quite a few outliers and/or isn't normally distributed. We can now save tidy and clean dataframe to the interim folder.

In [11]:
Dataset2.to_csv(r'data/01-interim/Single_chart_analyzer.csv', index=False)

### Data Cleaning: Single (Chart Analyzer)

Dataset 2 follows the same structure as Dataset 1 and upholds the same properties:

1. it has consistent column names

2. each row represents one chart

3. numeric columns are mostly numeric

4. there are no obvious formatting errors

5. the structure is already tidy

In terms of our project Dataset 2 does have more unnecessary variables we will have to filter out since this dataset is used merely for cross referencing arcade song presence.

### Animefest Player Scores (parsed.csv)

Animefest Player Scores is our first attempt to apply our model to a sample of new players during UCSD Animefest 2026. Our start time was 12:00pm and ended around 7:30pm. To take the data we used an online program of DDR called “Simply Love (v.5.1)” to mimic the arcade experience. From this data, the most relevant variables are the following:

Song_dir - What song they played

difficulty - uses their beginner/basic/difficult/expert labels

score - EX scores are given as EX x 1,000,000

The big concern during animefest was that we anticipated most people would like to play “anime” songs. Anime songs are usually custom charted and do not reflect chart complexity for standard DDR charts. Because we prioritize the well-being of our community, we would not interfere with song selection, but came to terms with the amount of valid data provided being significantly less than reported.


In [12]:
## YOUR CODE TO LOAD/CLEAN/TIDY/WRANGLE THE DATA GOES HERE
Dataset3 = pd.read_csv('data/00-raw/parsed.csv') # Read csv into dataframe
display(Dataset3)
Dataset3.isna().sum() #find categories with null values

Unnamed: 0                                           song_dir difficulty  \
0             0              Songs/DDR A20 PLUS/Midnight Amaretto/     Medium   
1             1              Songs/DDR A20 PLUS/Midnight Amaretto/     Medium   
2             2  Songs/Jubo 2022 OtakuVibes/KICK BACK [Chainsaw...     Medium   
3             3  Songs/Jubo 2022 OtakuVibes/KICK BACK [Chainsaw...     Medium   
4             4                            Songs/DDR A/Koishigure/     Medium   
..          ...                                                ...        ...   
192         192                 Songs/DDR WORLD/Funky Flux Fusion/       Hard   
193         193                  Songs/DDR WORLD/Gravity Collapse/       Easy   
194         194                  Songs/DDR WORLD/Gravity Collapse/  Challenge   
195         195                Songs/DDR WORLD/Time to HYPERDRIVE/       Hard   
196         196              Songs/DDR X2/roppongi EVOLVED ver. D/       Hard   

       steps_type  name   grade     score  percentdp  surviveseconds  \
0    dance-single   NaN  Failed  30024405     0.2871       78.376266   
1    dance-single   NaN  Tier13  79620344     0.6808      105.605774   
2    dance-single   NaN  Tier14  74654821     0.6400       92.431129   
3    dance-single   NaN  Tier12  85543873     0.7563       92.431129   
4    dance-single   NaN  Failed       915     0.0000       11.595005   
..            ...   ...     ...       ...        ...             ...   
192  dance-single   NaN  Tier11  85024485     0.7643      116.890854   
193  dance-single   NaN  Tier13  81490117     0.7173      119.538490   
194  dance-single   NaN  Failed    166935     0.0000       11.814010   
195  dance-single   NaN  Tier04  97810329     0.9707      120.111526   
196  dance-single   NaN  Tier05  97407450     0.9562      114.478973   

     maxcombo  ... radarvalues_chaos  radarvalues_notes  \
0          63  ...          0.287108              257.0   
1          65  ...          0.680836              274.0   
2          84  ...          0.640000              253.0   
3         125  ...          0.756364              263.0   
4          61  ...          0.000000              130.0   
..        ...  ...               ...                ...   
192       178  ...          0.764375              611.0   
193        44  ...          0.717347              168.0   
194       113  ...          0.000000              524.0   
195       423  ...          0.970794              632.0   
196       317  ...          0.956228              572.0   

    radarvalues_tapsandholds radarvalues_jumps radarvalues_holds  \
0                        236                19                22   
1                        252                22                22   
2                        247                 6                15   
3                        257                 6                15   
4                        117                13                16   
..                       ...               ...               ...   
192                      603                 8                21   
193                      160                 8                30   
194                      515                10                68   
195                      591                41                38   
196                      536                36                22   

    radarvalues_mines  radarvalues_hands  radarvalues_rolls  \
0                   0                  0                  0   
1                   0                  0                  0   
2                   0                  0                  0   
3                   0                  0                  0   
4                   0                  0                  0   
..                ...                ...                ...   
192                 0                  0                  0   
193                 0                  0                  0   
194                 0                  0                  0   
195        

Unnamed: 0                        0
song_dir                          0
difficulty                        0
steps_type                        0
name                            197
grade                             0
score                             0
percentdp                         0
surviveseconds                    0
maxcombo                          0
stageaward                      192
peakcomboaward                  197
modifiers                         0
datetime                          0
playerguid                        0
machineguid                       0
productid                         0
liferemainingseconds              0
disqualified                      0
pad                               0
stageguid                         0
guid                              0
tapnotescores_hitmine             0
tapnotescores_avoidmine           0
tapnotescores_checkpointmiss      0
tapnotescores_miss                0
tapnotescores_w5                  0
tapnotescores_w4            

As seen above this data is already tidy and in a usable format. The dataset looks tidy on the surface because it has consistent column names, each row represents one person, numeric columns are mostly numeric, there are no obvious formatting errors, and the structure is already tidy (one row per person, one column per variable). The dataset has 197 Entries and 49 columns of data. Of these 49 columns 3 have a majority NaN values "name", "stageaward", and "peakcomboaward". Because there isn't enough data for these rows to be useful they will be dropped.

In [13]:
Dataset3 = Dataset3.drop(columns=(["name", "stageaward", "peakcomboaward"]))
Dataset3.isna().sum().sum()

0

After cleaning there are 197 rows and 46 columns. We can now look for outliers.

In [14]:
df = Dataset3

# numeric-only for outlier detection
num = df.select_dtypes(include="number")

# tail cutoffs
lo = num.quantile(0.01)
hi = num.quantile(0.99)

# True where a value is in the extreme tails of its column
outlier_mask = num.lt(lo) | num.gt(hi)

# require at least k columns in the row to be flagged
k = 3 

row_outlier_counts = outlier_mask.sum(axis=1)
rows_flagged = row_outlier_counts >= k

print("Total rows:", len(df))
print(f"Rows flagged (>= {k} columns extreme):", rows_flagged.sum())

# show the worst rows first (most flagged columns)
worst_idx = row_outlier_counts.sort_values(ascending=False).head(50).index

# full set of flagged rows (all columns)
df_flagged = df.loc[rows_flagged].copy()
df.loc[worst_idx].assign(outlier_flag_count=row_outlier_counts.loc[worst_idx]).head(rows_flagged.sum())

Total rows: 197
Rows flagged (>= 3 columns extreme): 6


Unnamed: 0                                           song_dir difficulty  \
169         169              Songs/DDR X2/roppongi EVOLVED ver. D/       Hard   
195         195                Songs/DDR WORLD/Time to HYPERDRIVE/       Hard   
113         113               Songs/DDR WORLD/Ohisashi chuukagai!/   Beginner   
123         123  Songs/Touhou Fujinya ~ Touhou Pad Pack Rebirth...  Challenge   
145         145                    Songs/DDR A20 PLUS/ANTI ANTHEM/       Hard   
155         155                         Songs/DDR 6th Mix/MAX 300/     Medium   

       steps_type   grade     score  percentdp  surviveseconds  maxcombo  \
169  dance-single  Tier04  99384843     0.9786      114.974121       381   
195  dance-single  Tier04  97810329     0.9707      120.111526       423   
113  dance-single  Tier08  92807692     0.8871      115.818832        39   
123  dance-single  Failed  43693955     0.5233      104.157104       104   
145  dance-single  Failed   2447885     0.0240       36.287006        96   
155  dance-single  Tier04  97966649     0.9712       92.276604       277   

                                             modifiers  ... radarvalues_notes  \
169  4x, 53% Mini, FailImmediateContinue, Overhead,...  ...             573.0   
195  m690, 35% Mini, Reverse, FailImmediateContinue...  ...             632.0   
113  C250, 25% Mini, FailImmediateContinue, Overhea...  ...              39.0   
123  C690, 25% Mini, Reverse, FailImmediateContinue...  ...             618.0   
145  3.15x, 25% Mini, FailImmediateContinue, Overhe...  ...             530.0   
155  C455, 25% Mini, FailImmediateContinue, Overhea...  ...             385.0   

    radarvalues_tapsandholds radarvalues_jumps  radarvalues_holds  \
169                      537                36                 22   
195                      591                41                 38   
113                       39                 0                  0   
123                      608                10                 14   
145                      504                27                 75   
155                      372                13                 37   

     radarvalues_mines  radarvalues_hands  radarvalues_rolls  \
169                  0                  0                  0   
195                  0                  0                  0   
113                  0                  0                  0   
123                 87                  0                  0   
145                  0                  0                  0   
155                  0                  0                  0   

    radarvalues_lifts radarvalues_fakes  outlier_flag_count  
169                 0                 0                   6  
195                 0                 0                   4  
113                 0                 0                   3  
123                 0                 0                   3  
145                 0                 0                   3  
155                 0                 0                   3  

[6 rows x 47 columns]

Based on these results we can say that we may have a couple outliers in this dataset. Now we can send our cleaned and tidied dataset to the interim folder.

In [15]:
Dataset3.to_csv(r'data/01-interim/parsed-interim.csv', index=False)

### Data Cleaning: Animefest Player Scores

Current issues with the data (raw):

 The title includes additional variables such as song pack origin and adds clutter to the title. 

Numeric rating was missing.

EX score is scaled by 1,000,000


### Planned Merges

In the future we plan on developing a wrangled dataset using a combination of the important variables from dataset 1 and 2. The purpose of this dataset is to provide the us with a song encyclopedia devoid of songs outside the arcade version of DDR along with any other exclusives not accessible to new players. This will have the song TITLE of only (AC)(INTL) songs along with their corresponding DIFFICULTY X, Max EX%, median EX%, and amount of times the song was played. This dataset will be a unified dataset titled “DDR song Difficulty and Average Performance”. With this encyclopedia built we can make another unified data set that combines “DDR song Difficulty and Average Performance” to the “Animefest Player Scores”; assessing each player.

### NOTES:

At the time of the checkpoint, we have agreed the bias and skew of the data in dataset 1 that we will be computing median scores across all difficulty X's to serve as the baseline to better represent the new player demographic. and will work toward these developments and revisions for the EDA Checkpoint. 

## Ethics

Instructions: Keep the contents of this cell. For each item on the checklist
-  put an X there if you've considered the item
-  IF THE ITEM IS RELEVANT place a short paragraph after the checklist item discussing the issue.
  
Items on this checklist are meant to provoke discussion among good-faith actors who take their ethical responsibilities seriously. Your teams will document these discussions and decisions for posterity using this section.  You don't have to solve these problems, you just have to acknowledge any potential harm no matter how unlikely.

Here is a [list of real world examples](https://deon.drivendata.org/examples/) for each item in the checklist that can refer to.

[![Deon badge](https://img.shields.io/badge/ethics%20checklist-deon-brightgreen.svg?style=popout-square)](http://deon.drivendata.org/)

### A. Data Collection
 - [X] **A.1 Informed consent**: If there are human subjects, have they given informed consent, where subjects affirmatively opt-in and have a clear understanding of the data uses to which they consent?

    We will be collecting data from AnimeFest, on February 14th using a DDR pad provided by the DRGC club on campus. We will brief attendees of our project, how their data is used, and that it is completely optional. We won’t be collecting signatures as it is already made anonymous by the software.

 - [X] **A.2 Collection bias**: Have we considered sources of bias that could be introduced during data collection and survey design and taken steps to mitigate those?

    One form of collection bias we will encounter is the population we are sampling at AnimeFest. The majority of participants will most likely be UCSD students. Furthermore, we will be sampling from a small subset of UCSD students, those who are attending AnimeFest. However, we believe this bias will be negligible due to our other datasets being of varying age groups. Also, our goal is to collect a sample size of fifty to normalize the distribution.

 - [X] **A.3 Limit PII exposure**: Have we considered ways to minimize exposure of personally identifiable information (PII) for example through anonymization or not collecting information that isn't relevant for analysis?

     PII will be minimized at AnimeFest since the software ensures anonymity.

 - [X] **A.4 Downstream bias mitigation**: Have we considered ways to enable testing downstream results for biased outcomes (e.g., collecting data on protected group status like race or gender)?

   Addressed downstream bias in constraint and limitations of community-driven averages (how we got avg EX%) and how this may be an unfair comparison.  At animefest, we tried to keep players on the same settings so we can evaluate them equally, but this also opens the constraint of players having different preferences when playing rhythm games. When evaluating player performance, we must reason with if their score is in the scope of the avg EX%. One way we devised was calculating the IQR across difficulty X’s.

### B. Data Storage
 - [X] **B.1 Data security**: Do we have a plan to protect and secure data (e.g., encryption at rest and in transit, access controls on internal users and third parties, access logs, and up-to-date software)?
 - [X] **B.2 Right to be forgotten**: Do we have a mechanism through which an individual can request their personal information be removed?
 - [X] **B.3 Data retention plan**: Is there a schedule or plan to delete the data after it is no longer needed?
       
    No, one of our references is a “living” document of all player data and serves as a valuable resource to the rhythm game community.


### C. Analysis
 - [X] **C.1 Missing perspectives**: Have we sought to address blindspots in the analysis through engagement with relevant stakeholders (e.g., checking assumptions and discussing implications with affected communities and subject matter experts)?

    Relevant stakeholders include the high level/ optimization level players that are also accounted for in our dataset. This is something to be conscious about when looking at the scores relative to a predicted score of a new player. We have established a target demographic as the subject of interest in our rhythm game community. One other lingering concern would be DDR quality across all players. Some places do not take good care of their DDR machines and this leads to game impact (low sensitivity = losing score) and we acknowledge how this may affect averages.

 - [X] **C.2 Dataset bias**: Have we examined the data for possible sources of bias and taken steps to mitigate or address these biases (e.g., stereotype perpetuation, confirmation bias, imbalanced classes, or omitted confounding variables)?

    One possible bias that we encounter will be in our dataset that takes in registered users. A possibility is that with a highly-motivated player base like DDR, there will be players who work on optimization of their scores and might play lower level difficulties for accuracy; these players will bring the average score up. This bias is mitigated by the large sample size.

 - [X] **C.3 Honest representation**: Are our visualizations, summary statistics, and reports designed to honestly represent the underlying data?
 - [X] **C.4 Privacy in analysis**: Have we ensured that data with PII are not used or displayed unless necessary for the analysis?

    Yes. All player data involved in calculations are stripped independently from the individual who produced them and the players remain anonymous. At most, there is a tie to their in-game username, but that can be excluded as well.

 - [X] **C.5 Auditability**: Is the process of generating the analysis well documented and reproducible if we discover issues in the future?

     One of our resources is part of a living document that dates back almost 3 years ago.

### D. Modeling
 - [X] **D.1 Proxy discrimination**: Have we ensured that the model does not rely on variables or proxies for variables that are unfairly discriminatory?

   Our dataset does not include protected attributes such as race, gender, or age, so inherently our analysis cannot discriminate upon these characteristics. To avoid proxy discrimination we use DDR’s performance-based metrics that treats an individual as equal among other players such as difficulty X, EX, EX% and avg EX%.

   
 - [X] **D.2 Fairness across groups**: Have we tested model results for fairness with respect to different affected groups (e.g., tested for disparate error rates)?
 - [X] **D.3 Metric selection**: Have we considered the effects of optimizing for our defined metrics and considered additional metrics?

    For metrics we have developed 4 categories of data collection that all work in tandem with each other.
       
 - [X] **D.4 Explainability**: Can we explain in understandable terms a decision the model made in cases where a justification is needed?
 - [X] **D.5 Communicate limitations**: Have we communicated the shortcomings, limitations, and biases of the model to relevant stakeholders in ways that can be generally understood?

     We have communicated that our purpose is to appeal to the beginner/casual approach to starting rhythm games and the model will serve as a way to accommodate these members of our rhythm game community. Knowledge acquired through this data science may not serve a higher skill level, rather help reach out to those trying to get there.


### E. Deployment
 - [X] **E.1 Monitoring and evaluation**: Do we have a clear plan to monitor the model and its impacts after it is deployed (e.g., performance monitoring, regular audit of sample predictions, human review of high-stakes decisions, reviewing downstream impacts of errors or low-confidence decisions, testing for concept drift)?
 - [X] **E.2 Redress**: Have we discussed with our organization a plan for response if users are harmed by the results (e.g., how does the data science team evaluate these cases and update analysis and models to prevent future harm)?
 - [X] **E.3 Roll back**: Is there a way to turn off or roll back the model in production if necessary?
 - [X] **E.4 Unintended use**: Have we taken steps to identify and prevent unintended uses and abuse of the model and do we have a plan to monitor these once the model is deployed?
   
   Our model is designed to help new DDR players understand difficulty ranges through the numeric rating, but we understand using data geared toward the DDR community can result in skewed data that represents the competitive DDR scene via registered users. However, as a community that strives for improvement, these players unintentionally set a bar for players coming in; a bar that appears daunting at first. We will try our best to still represent our target demographic of new players while acknowledging long-time players.

## Team Expectations 

Team Expectation 1: Check Discord for Updates

Team Expectation 2: Be transparent on scheduling/conflicts

Team Expectation 3: If there are parts unclear about the project, bring it up as soon as possible


## Project Timeline Proposal

|Week|Main Task|Due Date| 
|:---:|---|:---:|
|5 (2/4)| Edit, finalize, and submit proposal<br>Search for datasets|Proposal Due (2/4)| 
|6 (2/11)| Data Checkpoint:<br>• Animefest this weekend, acquire sample<br>• Wrangle data of references<br>• Logistics as specified in proposal<br>• Make sure we have a way of obtaining necessary variables<br>| N/A|
|7 (2/18)|Edit, finalize, and submit Data checkpoint|Data Checkpoint Revisions (2/18)|
|8 (2/25)|EDA checkpoint:<br>• Substantive analysis<br>• Multiple plots<br>• Summary statistics<br>• Early insights<br>• Potential issues (outliers, skew, missingness) |N/A|
|9 (3/4)|Edit, finalize, and submit EDA checkpoint|EDA checkpoint Revisions (3/4)|
|10 (3/11)|Video|N/A|
|11 (3/18)|Edit, finalize, and submit Project|Whole Project + Video|